omnipaxos_core/src/storage.rs

```diff
@@ -355,52 +442,83 @@ where
         }
     }
 
+    fn load_cache(&mut self) {
+        self.state_cache.set_decided_idx(self.storage.get_decided_idx());
+        self.state_cache.set_accepted_round(self.storage.get_accepted_round());
+        self.state_cache.set_promise(self.storage.get_promise());
+        self.state_cache.set_compacted_idx(self.storage.get_compacted_idx());
+    }
+
     /*** Writing ***/
+    // Append entry, if the batch size is reached, flush the batch and return the actual
+    // accepted index (not including the batched entries)
     pub(crate) fn append_entry(&mut self, entry: T) -> u64 {
-        self.storage.append_entry(entry) + self.storage.get_compacted_idx()
+        self.state_cache.append_entry(entry);
+        if self.state_cache.get_batching_size() >= self.batch_size as u64 {
+            self.flush_batch();
+        }
+        self.get_log_len()
+    }
+
+    // Append entries in batch, if the batch size is reached, flush the batch and return the
+    // accepted index
+    pub(crate) fn append_entries(&mut self, entries: Vec<T>) -> Option<u64> {
```

Comment:
There are some cases where we want to actually append everything directly without batching (i.e., when synchronizing in the Prepare phase), so I suggest we add a parameter `skip_batching`:
```suggestion
    pub(crate) fn append_entries(&mut self, entries: Vec<T>, skip_batching: bool) -> Option<u64> {  
     // flush directly if skip_batching is true
```